File: 2020/day4/solution.js
const getInput = () => require('fs').readFileSync('./input.txt', 'utf8');

const validator = {
  "byr": (str) => 1920 <= str && str <= 2002,
  "iyr": (str) => 2010 <= str && str <= 2020,
  "eyr": (str) => 2020 <= str && str <= 2030,
  "hgt": (str) => {
    let valid = str.match(/^\d+(cm|in)$/) ?? false;
    let num = +str.slice(0, -2);
    if (!valid) return false; // return false if field contains invalid information
    return valid[1] == "in" ? 59 <= num && num <= 76 : 150 <= num && num <= 193
  },
  "hcl": (str) => /^#[0-9a-f]{6}$/.test(str),
  "ecl": (str) => /^(amb|blu|brn|gry|grn|hzl|oth)$/.test(str),
  "pid": (str) => /^[0-9]{9}$/.test(str)
};

function part1(passports) {
  return passports
    .filter((passport) => 
      Object.keys(validator)
        .every((key) => passport[key] != undefined)
    ).length;
}

function part2(passports) {
  return passports
    .filter((passport) => 
      Object.keys(validator)
        .every((key) => passport[key] != undefined && validator[key](passport[key]))
    ).length;
}

const passports = getInput().split("\n\n").map((s) => {
  let obj = {};
  for (let [key, value] of s.split(/\s|\n/).map((e) => e.split(":"))) {
    obj[key] = value;
  }
  return obj;
}); /// split, and trim space, convert string to struct
console.time();

console.log(`Number of passports: ${passports.length}`);
console.log(`part 1: ${part1(passports)}`);
console.log(`part 2: ${part2(passports)}`);
console.timeEnd();

File: 2020/day4/solution.py
# Source: https://www.reddit.com/r/adventofcode/comments/k6e8sw/2020_day_04_solutions/

import re

with open('input.txt', 'r') as fd:
    content = fd.read()
    lines = content.split("\n\n")
    input = [line.replace("\n", " ") for line in lines]

# Part 1
ValidTerms = ["byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid"]
ValidPassports = 0
for passport in input:
    if all(x in passport for x in ValidTerms):
        ValidPassports += 1

print("Part 1: The amount of valid passports is: " + str(ValidPassports))

# Part 2
ValidPassports = 0
for passport in input:
    if (re.search(r"byr:19[2-9]\d|byr:200[0-2]", passport) and re.search(r"eyr:202\d|eyr:2030", passport) and
            re.search(r"iyr:201\d|iyr:2020", passport) and re.search(r"hgt:1[5-8]\dcm|hgt:19[0-3]cm|hgt:59in|hgt:6\din|hgt:7[0-6]in", passport) and
    re.search(r"hcl:#[a-z0-9]{6}", passport) and re.search(r"ecl:(amb|blu|brn|gry|grn|hzl|oth)", passport) and re.search(r"pid:\d{9}\b", passport)):
        ValidPassports += 1

print("Part 2: The amount of valid passports is: " + str(ValidPassports))